Lazy loading only required exports from dynamic import()

pushkar100 opened the pull request:
PR for the feature request #10959 (resolves #10959)

Adding magic comments to lazy load only named exports from a dynamically imported module.
Introducing `webpackExports: "<name>"` or `webpackExports: ["<name1>", "<name2>"]`

At present, the entire module is exported. With this feature, we can reduce the code we import (i.e chunk files of smaller size are created)

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
Yes (will add more when required)

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**
- Addition to Webpack magic comments
- Benefit of using webpackExports

sokra commented on lib/dependencies/ImportParserPlugin.js:
```diff
@@ -149,6 +150,32 @@ class ImportParserPlugin {
 				}
+				if (importOptions.webpackExports !== undefined) {
+					if (
+						!importOptions.webpackExports ||
+						!(
+							importOptions.webpackExports.constructor.name === "String" ||
+							(importOptions.webpackExports.constructor.name === "Array" &&
+								importOptions.webpackExports.every(
+									item => item.constructor.name === "String"
+								))
+						)
```
```suggestion
						!(
							typeof importOptions.webpackExports === "string" ||
							(Array.isArray(importOptions.webpackExports) &&
								importOptions.webpackExports.every(
									item => typeof item === "string"
								))
						)
```

sokra commented on lib/dependencies/ImportParserPlugin.js:
```diff
@@ -149,6 +150,32 @@ class ImportParserPlugin {
+						);
+					} else {
+						// Convert export names to be of type string[][]:
+						if (importOptions.webpackExports.constructor.name === "String") {
```
typeof

evilebottnawi commented:
I think supporting `import { something } from /* webpackExports: "something" */ 'package';` should be good feature too

evilebottnawi commented:
@pushkar100 What about if I want to export something as default - `import(webpackExports: "customFunction").then((myFunction) => myFunction())`?

pushkar100 commented:
@evilebottnawi I did think about it. The `import()` proposal is supposed to return a promise for the module namespace object which contains all the imports including a "default" import (Please correct me if I am wrong). Won't we be creating an alternative pattern by resolving to the default export instead of the module object itself? 
The idea I currently have for accessing default export is:
```javascript
import(/* webpackExports: "default" */ "./module")
.then(module => /* only module.default is available */)
```

sokra commented:
The code should keep working even if the comment is ignored (e.g. in native esm). So even if you only request a single export please return a namespace object for compat with native esm.

sokra commented:
One can use destructing for shorter syntax: `import("module").then(({ default: theDefaultExport }) => doSomething(theDefaultExport))` or `const { default: theDefaultExport } = await import("module")`

pushkar100 commented:
@sokra @evilebottnawi  I have prevented mangling of exports when `webpackExports` is used.
Also, added basic tests for checking unmangled names and default exports.

Now we can use `webpackExports: "default"` to get only the default export.

sokra commented on lib/Dependency.js:
```diff
@@ -62,6 +62,8 @@ class Dependency {
 		/** @type {DependencyLocation} */
 		this.loc = undefined;
+		/** @type {string[][] | null} */
+		this.referencedExportNames = null;
```
This should not be here. Every info should be passed via `getReferencedExports` and not via magic properties.

sokra commented on lib/dependencies/ImportWeakDependency.js:
```diff
@@ -5,24 +5,37 @@
+	/**
+	 * Returns list of exports referenced by this dependency
+	 * @param {ModuleGraph} moduleGraph module graph
+	 * @returns {string[][]} referenced exports
```
I think the return type need to be changed to support passing `canMangle` which then sets `canMangleUsed`

sokra commented on test/cases/chunks/inline-options/index.js:
```diff
@@ -81,54 +89,141 @@ it("should be able to use weak mode (without context)", function() {
+		})
+		.catch(function (e) {
+			expect(e).toMatchObject({
+				message: /not available/,
```
There should not be an error, otherwise your checks above didn't run. Make sure the module is available, e.g. by importing something from it.

sokra commented on test/cases/chunks/inline-options/index.js:
```diff
@@ -81,54 +89,141 @@ it("should be able to use weak mode (without context)", function() {
+	});
+});
+
+it("should not contain only default from webpackExports from module", function () {
```
There is a not to much in the name

sokra commented on lib/FlagDependencyUsagePlugin.js:
```diff
@@ -132,7 +140,19 @@ class FlagDependencyUsagePlugin {
-						processModule(connection.module, referencedExports);
+						let hasInlineWebpackExports = false;
+						if (
+							dep.referencedExportNames &&
```
Such info should be passed via `getReferencedExports`

pushkar100 commented:
@sokra I have refactored the code: Using an object for referenced exports with canMangle option.
Fixed the basic test cases (weak mode & removal of "not" from title).

sokra commented on lib/FlagDependencyUsagePlugin.js:
```diff
@@ -55,9 +56,15 @@ class FlagDependencyUsagePlugin {
-											exportName
+											exportName,
+											canMangle
 										);
```
Thats a weird way to set canMangleUse in the ModuleGraph with a `get...` method.
`exportInfo.canMangleUse = false;`

``` js
const exportItem = usedExport[i];
let exportInfo;
if (typeof exportItem === "object") {
  exportInfo = currentExportsInfo.getExportInfo(exportItem.name);
  if (exportInfo.canMangle === false) exportInfo.canMangleUse = false;
} else {
  exportInfo = currentExportsInfo.getExportInfo(exportItem);
}
```

pushkar100 replied:
After invoking `getExportInfo`, we are getting `exportInfo.canMangle` to be `true`. Can I change the `if` condition to the following: `if (exportItem.canMangle === false) exportInfo.canMangleUse = false;`?

sokra replied:
Oh yes that was a typo

pushkar100 replied:
okay :)

sokra commented on lib/ModuleGraph.js:
```diff
@@ -211,15 +211,22 @@ class ExportsInfo {
 	 * @returns {ExportInfo} export info for this name
 	 */
-	getExportInfo(name) {
+	getExportInfo(name, canMangle) {
```
Revert that

pushkar100 replied:
Will do

sokra commented on lib/dependencies/ImportWeakDependency.js:
```diff
@@ -5,24 +5,35 @@
+	constructor(request, range, names) {
 		super(request);
-
+		this.names = names;
```
```suggestion
		this.referencedExports = referencedExports;
```

sokra commented on types.d.ts:
```diff
@@ -1683,7 +1683,7 @@ declare class Dependency {
 	 * Returns list of exports referenced by this dependency
 	 */
-	getReferencedExports(moduleGraph: ModuleGraph): string[][];
+	getReferencedExports(moduleGraph: ModuleGraph): any[][] | string[][];
```
This should be a more specific type

pushkar100 replied:
will make the change

sokra commented on types.d.ts:
```diff
@@ -2301,7 +2301,7 @@ declare class ExportsInfo {
 	setHasUseInfo(): void;
 	getOwnExportInfo(name: string): ExportInfo;
-	getExportInfo(name: string): ExportInfo;
+	getExportInfo(name: string, canMangle?: boolean): ExportInfo;
```
This should not change.

pushkar100 replied:
Will make the change

pushkar100 commented on lib/Dependency.js:
```diff
@@ -93,7 +93,7 @@ class Dependency {
 	 * Returns list of exports referenced by this dependency
 	 * @param {ModuleGraph} moduleGraph module graph
-	 * @returns {string[][]} referenced exports
+	 * @returns {string[][] | Object[][]} referenced exports
```
Is this okay?

```javascript
/**
 * @typedef {Object} ReferencedExport
 * @property {string} name
 * @property {boolean=} canMangle
 */
```

```javascript
/**
* Returns list of exports referenced by this dependency
* @param {ModuleGraph} moduleGraph module graph
* @returns {(string[] | ReferencedExport[])[]} referenced exports
*/
getReferencedExports(moduleGraph) { }
```

sokra replied:
That would work, but let's save some objects and combine it a little bit:

`property​ {​string[]} name`

`(string[] | ReferencedExport)[]`

pushkar100 replied:
Okay, I will define an array containing the object as `ReferencedExport`

pushkar100 replied:
I was not able to find the most elegant way to define an element inside an array in jsdocs, this is what I intend to use:
```javascript
/**
 * @typedef {Object} ReferencedExportItem
 * @property {string} name name of the referenced export
 * @property {boolean=} canMangle when true, referenced export can be mangled
 */

/** @typedef {ReferencedExportItem[]} ReferencedExport */
```
Hoping this is fine @sokra

pushkar100 commented:
@sokra I've incorporated the type and refactoring changes. 

An additional change in `WasmFinalizeExportsPlugin.js` was required so that `jsIncompatibleExports[name]` does not throw type related error on receiving an object

sokra commented:
Thanks
